### PR TITLE
User#can_edit? fix

### PR DESCRIPTION
**Pull request by TunrayoIlawole**
## What this PR does
This PR addresses #5948 by ensuring users with multiple roles are correctly assigned editing permissions.

## More details on what was done

- Updated User#role method to properly handle cases where a user has multiple roles in a course.
- The method now checks if the user has any editing role and returns the appropriate role.
- Added unit test to verify that users with both campus volunteer (non-editing role) and instructor roles are correctly recognized as having editing privileges.

### CODE REVIEW

**Comment by ragesoss** [comment]
I think this would be better solved by implementing a different method to check for which roles a user has for a course. The User#role method is designed (and named) to return a single user role, but for can_edit? we want to consider all the roles a user has for a course, and check whether any of them are editing roles.

Maybe something like this would work in place of the call to `User#role`: `return true if CoursesUsers.exists?(course: course, user: self, role: EDITING_ROLES)`

**Comment by TunrayoIlawole** [comment]
Thank you @ragesoss . Noted, i will make the update

**Comment by TunrayoIlawole** [comment]
Hi, @ragesoss . I created the new method to check whether any of the roles a user has for a course is an editing role and called that in `can_edit`

**Comment by ragesoss** [comment]
Please check the build log; it looks like this has broken something.

**Comment by TunrayoIlawole** [comment]
Hi @ragesoss, I noticed that in cases where the course is nill, the call to User#role returns an Instructor role which ultimately returns a true value, bypassing the need to check User#campaign_organizer?

For my new method, should i follow a similar approach and return true if the course is nill, since an instructor role is considered an editing role? Or would it be more appropriate to implement a check that returns early in User#campaign_organizer?

I’d appreciate your advice on the best way to handle this case to ensure it aligns with the project's standards. Thank you.

**Comment by ragesoss** [comment]
The behavior of returning the instructor role for cases where the course is nil is confusing, and I'm not exactly sure how we use that. Based on the comment, it seems to be related to the course creation process, before the CoursesUsers record for the instructor has been created, because the course itself as not been saved in the database. But the best thing to do might be to identify exactly where we use that behavior (for example, by throwing an error instead of returning the instructor role, and seeing which tests fail) and handle that differently, so that `role` can be a little more intuitive.

**Comment by TunrayoIlawole** [comment]
Hi, @ragesoss, thank you for your feedback. The behaviour is not really relied on, except in determining whether a user has an editing role. 

The main issue with assigning an Instructor role to a user that is not associated with any course (i.e course is nil) is that it automatically gives the user an editing role. This prevented the error-prone User#campaign_organizer?, which currently does not handle nil values for course and therefore fails, from being called.

This is the code that is failing. `course.campaigns` throws an error because course is nil
```
  def campaign_organizer?(course)
    CampaignsUsers.exists?(user: self, campaign: course.campaigns,
                           role: CampaignsUsers::Roles::ORGANIZER_ROLE)
  end
```

The pipeline is failing because User#campaign_organizer? is now being called as a result of the changes i made to the User#can_edit code to no longer rely on User#role for determining whether a user has an editing role or not.

I propose:
- Returning false in User#campaign_organizer? if the course is nill
- Assigning a default, non-editing role to a user that is not associated with a course or raising an error . Either of these methods will require significant changes to the logic and tests.

Please let me know what you think, @ragesoss

**Comment by ragesoss** [comment]
Which tests fail when you throw an error from the course.nil? check?

**Comment by TunrayoIlawole** [comment]
Hi, @ragesoss , the following tests fail when i throw an error from the nil check:

- [User#role grants instructor permission for a user creating a new course](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/models/user_spec.rb#L55)
- [AssignmentsController PATCH #update_status when a status param is provided renders a 200](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/controllers/assignments_controller_spec.rb#L422)
- [AssignmentsController PATCH #update_status when no status param is provided renders a 422](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/controllers/assignments_controller_spec.rb#L432)
- [AssignmentsController PATCH #update_sanbox_url updating sandbox url with valid urls update sandbox url successfully with example 1](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/controllers/assignments_controller_spec.rb#L452)
- [AssignmentsController PATCH #update_sanbox_url updating sandbox url with valid urls update sandbox url successfully with example 2](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/controllers/assignments_controller_spec.rb#L459)
- [AssignmentsController PATCH #update_sanbox_url updating sandbox url with valid format but belongs to different wiki does not update url and send response with status: unprocessable entity](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/controllers/assignments_controller_spec.rb#L475)
- [AssignmentsController PATCH #update_sanbox_url updating sandbox url with invalid url format does not update url and send response with status: bad request example 1](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/controllers/assignments_controller_spec.rb#L490)
- [AssignmentsController PATCH #update_sanbox_url updating sandbox url with invalid url format does not update url and send response with status: bad request example 2](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/controllers/assignments_controller_spec.rb#L498)


There are other tests that fail for both implementations(i.e with error thrown and with no error thrown), such as:
[updating locale via URL updates the locale and redirects to the home page](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/features/update_locale_spec.rb#L5)

and other errors from Capybara

**Comment by ragesoss** [comment]
Hmm... that's weird. I would not expect `course` to be `nil` in those AssignmentsController tests, as creating the course record is part of the setup. Did you configure it to only throw an error when `course` is `nil`, or every time that line is reached?

**Comment by TunrayoIlawole** [comment]
Hi @ragesoss , sorry i should have clarified. The AssignmentsController tests are failing because User#campaign_organizer? is now being called in the code. It was not called before because User#can_edit? automatically returned true as a result of the instructor role being returned from User#role.

So i belive the only test that is directly failing as a result of the error that i threw in User#role is [User#role grants instructor permission for a user creating a new course](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/models/user_spec.rb#L55)

**Comment by ragesoss** [comment]
Okay. I'm still not clear on why the assignment tests rely on the instructor role behavior. Is it because the test setup is incorrect?

**Comment by TunrayoIlawole** [comment]
They do not rely on the instructor role @ragesoss . Apologies for the confusion. They are failing because of a missing nil check in User#campaign_organizer. The reason why I kept referencing the AssignmentsController tests is because I wanted to give some context as to why they were not failing with the previous implementation.

**Comment by ragesoss** [comment]
Okay. So it sounds like, as suspected, the only test that directly relies on the instructor role-for-nil-course behavior is the one related to creating a new course, but that behavior was also masking either a bug with the campaign_organizer check, or a bug with test setup related to it. Before adding a new nil check, we want to understand whether any actual valid requests can have a nil value, or whether it's just something that is missing from the test setup but in actual usage it could never be nil.

The other thing to look into is how `role` gets called when creating a new course, as we may want to find a different way of handling it.

**Comment by TunrayoIlawole** [comment]
Hi @ragesoss , thank you for your feedback.

From my investigation, i can see that User#role is only called in the application to verify edit permissions and [confirm access to a course](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/app/controllers/courses_controller.rb#L424).

When a course is created by a user, that user is automatically assigned an instructor role. Also, the course is returned whether it was successfully persisted in the database or not. I believe this is the reason why User#role was set to return an instructor role for an empty course.

So the idea is whenever User#role (User = current user i.e the user that created the course) is called with a nil course value, it automatically returns the instructor role so as to avoid not getting any result [here](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/app/models/user.rb#L175) in case the course has not been saved to the database.

But technically, i do not think any actual valid requests can have a nil course value since a course object is returned whether the course was successfully persisted or not, so User#role will always be called with a course object.

`role` does not get called when creating a new course.

Also, i think the description of this [test](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/df2fac6f923e61f78282b0d3ceb67546cb169e7b/spec/models/user_spec.rb#L56) might be misleading because User#role only determines the role of a user in a given course, it has nothing to do with course creation.

Please let me know what you think.

**Comment by ragesoss** [comment]
Interesting. So it sounds like we should probably remove that behavior altogether, for handling a nil course.

I think implementation of `User#role` is very confusing, as it is really more about permissions than roles. Since it is only called from a few places, I think the best thing would be to refactor it from `#role` to `#course_roles`, make that method simply return an array roles the user has in the course, and move the logic for handling permissions for admins and campaign organizers over to the methods that are explicitly about permission (`User#can_edit?` and maybe a new `User#can_view?` for use in `protect_privacy`.)

**Comment by TunrayoIlawole** [comment]
Noted @ragesoss , I will work on that.

The AssignmentsController tests are also failing because for some reason, when User#can_edit is called in AssignmentsController#check_permissions, an empty course is passed. I am looking into that

**Comment by TunrayoIlawole** [comment]
Hi @ragesoss , I discovered that most of the AssignmentControllers tests that are failing are failing due to incomplete request parameters being passed. 

But i am still having an issue with two of the tests: [1](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/887d8da25c76b672366bdd30f9ccca4464fa095d/spec/controllers/assignments_controller_spec.rb#L422) & [2](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/887d8da25c76b672366bdd30f9ccca4464fa095d/spec/controllers/assignments_controller_spec.rb#L432) and it has to do with not calling `set_course` before executing `update_status` - [see code](https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/887d8da25c76b672366bdd30f9ccca4464fa095d/app/controllers/assignments_controller.rb#L12)

Is there a reason for that behaviour?

**Comment by ragesoss** [comment]
The JS action that hits the update_status endpoint does not include any course identifier, so I think set_course would fail. This comes from `updateAssignmentStatus` in the assignment_actions.js file.

**Comment by ragesoss** [comment]
Probably the right approach is to set the course via the assignment (as with `#destroy`) before performing the permissions check.

**Comment by ragesoss** [comment]
@TunrayoIlawole yes

**Comment by TunrayoIlawole** [comment]
Hi @ragesoss , here are all the changes I made to the code:

- created a method User#editing_role? to check if a user has any editing role in a course 
- created a method to return an array of all the courses a user has in a course
- created User#can_view method
- updated User#nonvisitor? to return true if a user has any role other than a visitor role
- updated User#can_see_real_names? to return true if a user has any role that is included in the `REAL_NAME_ROLES` object
- created tests for all new methods
- updated `request_params` in some assignment controller tests to include the `course_slug` so that the course can be set correctly via AssignmentsController#set_course
- set course before checking permissions in AssignmentsController#update_status and AssignmentsController#claim

File: app/controllers/assignments_controller.rb [hunk 1]
```diff
@@ -49,7 +49,7 @@ def assign_reviewers_randomly
 
   # Select an Available Article as a new Assignment for a user
   def claim
-    @claimed_assignment = Assignment.find(params[:assignment_id])
+    @claimed_assignment = Assignment.find(params[:id])
```
Comment (ragesoss) [inline app/controllers/assignments_controller.rb:52]: Did you run the tests locally? This is a change I would expect to break things, as the parameter for this action from the `routes.rb` code is passed as `assignment_id`.

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -76,36 +62,23 @@
...
     end
 
-    it 'returns -1 for a user who is not part of the course' do
-      course = create(:course)
-      user = create(:user)
-      role = user.role(course)
-      expect(role).to eq(-1)
+    it 'returns an empty array when the user has no roles in the course' do
+      course = create(:course,
+                      id: 1)
```
Comment (ragesoss) [inline spec/models/user_spec.rb:77]: We generally want to avoid setting `id` in artificial ways like this for tests. Especially when using the same value for different purposes, it can be confusing and also makes it easier to miss bugs.

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -140,6 +113,156 @@
...
+      create(:courses_user,
+             course_id: 1,
+             user_id: 1,
+             role: 1) # instructor
+
+      # User is also a campus volunteer
+      create(:courses_user,
+             course_id: 1,
+             user_id: 1,
+             role: 2) # campus volunteer
```
Comment (ragesoss) [inline spec/models/user_spec.rb:132]: Use the existing constants rather than magic numbers for this (which also eliminates the need for a comment).

**Comment by ragesoss** [comment]
I think that strategy makes sense. I would suggest naming these methods very clearly: have `#user_roles` return all roles, and have `#highest_role` return the highest role, and any additional functionality required for a specific reason should have its own method building on top of those (or implement additional logic independently).

My priority is to make the public methods on this widely-used class as simple and self-explanatory as possible.

File: app/controllers/courses_controller.rb [hunk 4]
```diff
@@ -421,7 +421,7 @@ def verify_edit_credentials
   def protect_privacy
     return unless @course.private
     # Admins and enrolled users have non-visitor roles
-    return if current_user && current_user.role(@course) != CoursesUsers::Roles::VISITOR_ROLE
+    return if current_user&.can_view?(@course)
```
Comment (ragesoss) [inline app/controllers/courses_controller.rb:424]: Why not just use `nonvisitor?`

File: app/models/user.rb [hunk 5]
```diff
@@ -163,31 +163,43 @@ def course_student?
...
-    role(course) != CoursesUsers::Roles::VISITOR_ROLE
+    return true if admin?
+    roles = course_roles(course)
+    roles.size > 1 || (roles.size == 1 && roles.first != CoursesUsers::Roles::VISITOR_ROLE)
+  end
+
+  # returns an array of roles a user has in a given course
+  def course_roles(course)
+    user_course_roles = course.courses_users.where(user_id: id).order('role DESC').pluck(:role)
+    return user_course_roles
```
Comment (ragesoss) [inline app/models/user.rb:174]: Better to make this a one line method without assigning it to a variable or using explicit return.

File: app/models/user.rb [hunk 6]
```diff
@@ -163,31 +163,43 @@ def course_student?
   end
 
   def nonvisitor?(course)
-    role(course) != CoursesUsers::Roles::VISITOR_ROLE
+    return true if admin?
+    roles = course_roles(course)
+    roles.size > 1 || (roles.size == 1 && roles.first != CoursesUsers::Roles::VISITOR_ROLE)
```
Comment (ragesoss) [inline app/models/user.rb:168]: The visitor role does not exist as a CoursesUsers record, so it should never be among the roles returned by `course_roles`

File: app/models/user.rb [hunk 7]
```diff
@@ -163,31 +163,43 @@ def course_student?
...
+    # Check if the user has an editing role in the course
+    CoursesUsers.exists?(course:, user: self, role: EDITING_ROLES)
   end
 
   EDITING_ROLES = [CoursesUsers::Roles::INSTRUCTOR_ROLE,
                    CoursesUsers::Roles::WIKI_ED_STAFF_ROLE].freeze
   def can_edit?(course)
     return true if admin?
-    return true if EDITING_ROLES.include? role(course)
+    return true if editing_role?(course)
```
Comment (ragesoss) [inline app/models/user.rb:194]: Since this concept of editing roles is only used here, I think a better option would be to take the intersection of course_roles and editing roles, like `(EDITING_ROLES & course_roles(course)).any?`, or the same `any?` pattern you are implementing in `can_see_real_names?`. No need for a separate method.

File: app/models/user.rb (hunk 6 above)
Comment (TunrayoIlawole) [reply app/models/user.rb:168]: You are right, @ragesoss . I guess a better approach will be to check that the `course_roles` array is not empty, right?

Comment (ragesoss) [reply app/models/user.rb:168]: yeah. course_role().empty? should work.

**Comment by ragesoss** [comment]
I suspect that these two failing tests are related to the changes; they seem like tests that would be affected by 'can edit' checks.

**Comment by TunrayoIlawole** [comment]
Thank you, @ragesoss . Update made

File: app/models/user.rb [hunk 8]
```diff
@@ -175,6 +175,7 @@ def course_roles(course)
   def highest_role(course)
     roles = course_roles(course)
 
+    return CoursesUsers::Roles::INSTRUCTOR_ROLE if admin?
```
Comment (ragesoss) [inline app/models/user.rb:178]: This should be done before the query for roles, as it may replace the need to make a database query.

File: app/models/user.rb [hunk 9]
```diff
@@ -163,31 +163,37 @@ def course_student?
...
                    CoursesUsers::Roles::WIKI_ED_STAFF_ROLE].freeze
   def can_edit?(course)
     return true if admin?
-    return true if EDITING_ROLES.include? role(course)
+    return true if course_roles(course).any? { |role| EDITING_ROLES.include?(role) }
     return true if campaign_organizer?(course)
     false
   end
 
+  def can_view?(course)
```
Comment (ragesoss) [inline app/models/user.rb:193]: This is not needed, correct? (The tests can just be changed to `nonvisitor?`.)

Comment (TunrayoIlawole) [reply app/models/user.rb]: Update made and tested @ragesoss
